modules/heat_conduction/src/actions/ThermalContactAction.C

```diff
@@ -0,0 +1,289 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ThermalContactAction.h"
+
+#include "AddVariableAction.h"
+#include "FEProblem.h"
+#include "libmesh/string_to_enum.h"
+#include "GapConductance.h"
+#include "NonlinearSystem.h"
+
+registerMooseAction("HeatConductionApp", ThermalContactAction, "add_aux_kernel");
+registerMooseAction("HeatConductionApp", ThermalContactAction, "add_aux_variable");
+registerMooseAction("HeatConductionApp", ThermalContactAction, "add_bc");
+registerMooseAction("HeatConductionApp", ThermalContactAction, "add_dirac_kernel");
+registerMooseAction("HeatConductionApp", ThermalContactAction, "add_material");
+registerMooseAction("HeatConductionApp", ThermalContactAction, "add_slave_flux_vector");
+
+template <>
+InputParameters
+validParams<ThermalContactAction>()
+{
+  InputParameters params = validParams<Action>();
+  params.addClassDescription(
+      "Action that controls the creation of all of the necessary objects for "
+      "calculation of Thermal Contact");
+
+  params.addParam<std::string>(
+      "gap_aux_type",
+      "GapValueAux",
+      "A string representing the Moose object that will be used for computing the gap size");
+  params.addRequiredParam<NonlinearVariableName>("variable", "The variable for thermal contact");
+  params.addRequiredParam<BoundaryName>("master", "The master surface");
+  params.addRequiredParam<BoundaryName>("slave", "The slave surface");
+  params.addParam<Real>("tangential_tolerance",
+                        "Tangential distance to extend edges of contact surfaces");
+  params.addParam<Real>(
+      "normal_smoothing_distance",
+      "Distance from edge in parametric coordinates over which to smooth contact normal");
+  params.addParam<std::string>("normal_smoothing_method",
+                               "Method to use to smooth normals (edge_based|nodal_normal_based)");
+
+  MooseEnum orders(AddVariableAction::getNonlinearVariableOrders());
+  params.addParam<MooseEnum>("order", orders, "The finite element order");
+
+  params.addParam<bool>(
+      "warnings", false, "Whether to output warning messages concerning nodes not being found");
+  params.addParam<bool>(
+      "quadrature", false, "Whether or not to use quadrature point based gap heat transfer");
+
+  params.addParam<std::string>(
+      "appended_property_name", "", "Name appended to material properties to make them unique");
+  params.addRequiredParam<std::string>(
+      "type",
+      "A string representing the Moose object that will be used for heat conduction over the gap");
```

Comment:
I know the is how it has been for a long time but the use of this type parameter causes issues with the creation of the documentation (not all parameters appear properly on the pages) if I remember my discussion with @dschwen before.  It also messes with the auto completion in atom because it uses a similar format as the documentation system to read in information.  

As far as I know this type is for the model to use in gap heat transfer (i.e., GapHeatTransfer or GapHeatTransferLWR) and is used to trigger the correct creation of either GapConductance or GapConductanceLWR respectively.  Normally actions do not have a type parameter so could we change this parameter to be something else and get it to work like `gap_model`. 

I do like that it is being refactored into a single action and I think some of these cleanups/modifications should occur at the same time if possible.

Reply:
Interesting point, makes sense it should be something other than `type`. That would take some serious deprecation though right? Maybe it's worth waiting until the entire thing is rewritten as a constraint?

Reply:
It would need definitely need some serious deprecation.  I do like this first step of combining everything into the single action (my preferred method of reading through or developing actions).  We could a similar deprecation on this parameter as was done for the displacements (as you've noted we still haven't  removed the old way). This can be done in another PR.